File: skills/install-slop-guard/assets/python/rules/smells/no_lazy_class.py
"""Reject classes that do not earn their keep."""

from __future__ import annotations

import ast

from ...rule import Rule

EXEMPT_BASES = frozenset({"Protocol", "ABC", "Enum", "IntEnum", "StrEnum", "Exception", "TypedDict", "NamedTuple"})


class NoLazyClass(Rule):
    name = "no-lazy-class"
    group = "smells"
    reference = "https://refactoring.guru/smells/lazy-class"
    description = "Disallow a class with no state and at most one method; a function says the same thing."

    def visit_ClassDef(self, node: ast.ClassDef) -> None:
        if node.decorator_list or any(
            self.unparse(base).split("[")[0].split(".")[-1] in EXEMPT_BASES for base in node.bases
        ):
            self.generic_visit(node)
            return
        methods = [
            statement
            for statement in node.body
            if isinstance(statement, (ast.FunctionDef, ast.AsyncFunctionDef))
            and not (statement.name.startswith("__") and statement.name.endswith("__"))
        ]
        has_state = any(
            isinstance(child, ast.Attribute)
            and isinstance(child.ctx, ast.Store)
            and isinstance(child.value, ast.Name)
            and child.value.id == "self"
            for child in ast.walk(node)
        )
        fields = [statement for statement in node.body if isinstance(statement, (ast.Assign, ast.AnnAssign))]
        if not has_state and not fields and len(methods) <= 1 and methods:
            self.report(
                node,
                f"`{node.name}` holds no state and has one method, so the class adds a name and a "
                f"call without adding a decision. Make `{methods[0].name}` a function.",
            )
        self.generic_visit(node)

File: languages/typescript/src/smells/rules/no-lazy-class.ts
import { defineRule } from "@oxlint/plugins";

import { walk } from "../shared/traversal.ts";

/** Ban classes that do not earn their keep. */
export const noLazyClassRule = defineRule({
	meta: {
		type: "problem",
		docs: {
			description:
				"Disallow a class with no fields and at most one method; a function says the same thing with less ceremony.",
		},
		messages: {
			lazyClass:
				"`{{name}}` holds no state and has one method, so the class adds a name and a call without adding a decision. Make `{{method}}` a function.",
		},
	},
	createOnce(context) {
		return {
			ClassBody(node) {
				const parent = node.parent;
				if (parent?.type !== "ClassDeclaration" && parent?.type !== "ClassExpression") return;
				if (parent.superClass !== null && parent.superClass !== undefined) return;
				if (parent.abstract === true || (parent.decorators?.length ?? 0) > 0) return;
				const fields = node.body.filter(
					(member) => member.type === "PropertyDefinition" || member.type === "AccessorProperty",
				);
				const methods = node.body.filter(
					(member) => member.type === "MethodDefinition" && member.kind === "method",
				);
				const constructors = node.body.filter(
					(member) => member.type === "MethodDefinition" && member.kind === "constructor",
				);
				if (fields.length > 0 || methods.length !== 1 || constructors.length > 0) return;
				let holdsState = false;
				walk(node, context.sourceCode.visitorKeys, (child) => {
					if (
						child.type === "AssignmentExpression" &&
						child.left.type === "MemberExpression" &&
						child.left.object.type === "ThisExpression"
					) {
						holdsState = true;
					}
				});
				if (holdsState) return;
				const only = methods[0];
				context.report({
					node,
					messageId: "lazyClass",
					data: {
						name: parent.id?.name ?? "This class",
						method:
							only !== undefined && only.type === "MethodDefinition" && only.key.type === "Identifier"
								? only.key.name
								: "it",
					},
				});
			},
		};
	},
});
